### PR TITLE
feat: add metascraper-logo-bimi

### PR DESCRIPTION
**Pull request by Kikobeats**
Resolves `logo` from the [BIMI](https://datatracker.ietf.org/doc/html/draft-blank-ietf-bimi) DNS record, the standard behind the brand logo mailbox providers show next to an email.

```
$ dig +short TXT default._bimi.microlink.io
"v=BIMI1; l=https://cdn.microlink.io/logo/logo.svg;"
```

## Why

The spec constrains the logo to SVG Tiny P/S, so every hit is vector, square and transparent:

```
x.com          viewBox="0 0 400 400"       baseProfile="tiny-ps"  <title>X Corp.</title>
apple.com      viewBox="0 0 800 800"       baseProfile="tiny-ps"  <title>Apple</title>
adobe.com      viewBox="0 0 810.48 810.48" baseProfile="tiny-ps"  <title>Adobe: ...</title>
cloudflare.com viewBox="0 0 200 200"       baseProfile="tiny-ps"  <title>Cloudflare Inc.</title>
```

38 of the 47 brand hits below also publish a Verified Mark Certificate (`a=`), meaning a CA verified the mark against the trademark owner. It needs no markup either, so it resolves on JavaScript rendered or unreachable pages, and a TXT lookup measured 5-8ms warm against a HTTP round trip in `metascraper-logo-favicon`.

## Coverage

`dig TXT default._bimi.<domain>` over 124 domains:

| set | with an `l=` logo |
|---|---|
| Top 100 brands | 47/100 |
| domains used across this repo's tests | 4/24 (incl. `microlink.io`) |

High precision, low recall. That's why it ships as its own package rather than folded into `metascraper-logo-favicon`: its provider list runs after the DOM icon scan, and every BIMI publishing brand also ships an `apple-touch-icon`, so it would never fire there. As a separate bundle `mergeRules` respects declaration order and the caller decides:

```js
const metascraper = require('metascraper')([
  require('metascraper-logo-bimi')(),
  require('metascraper-logo')(),
  require('metascraper-logo-favicon')()
])
```

Nothing existing changes behavior unless the package is added.

## Details

- Lookup uses the registrable domain, so `https://blog.cloudflare.com/post` resolves `default._bimi.cloudflare.com`.
- Memoized per domain via `@keyvhq/memoize`, including misses.
- Declination records (`v=BIMI1; l=;`) are honored as an explicit opt out.
- Non `https` locations are rejected, per spec.
- The URL is verified reachable and served as `image/svg+xml` before being returned.
- `resolveTxt` is injectable for runtimes without `node:dns`; the README documents a DNS over HTTPS resolver, verified working against Cloudflare.
- The record is self asserted DNS, so trust is equivalent to `og:logo`. The VMC is not parsed.

## Tests

32 tests, all green. Record parsing, DNS to logo resolution, reachability and content type, rule integration, and memoization are covered with stubbed resolvers; `test/dns.js` exercises the default system resolver against a live record.

```
32 tests passed
```

End to end against real domains:

```
https://blog.cloudflare.com/a-post   https://www.cloudflare.com/cloudflare_1171114652.svg
https://www.apple.com/               https://www.apple.com/bimi/v2/apple.svg
https://vercel.com/docs              https://vercel.com/static/bimi.svg
https://github.com/microlinkhq       null
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01VYx23URhjLXNWiUELZqPoq

---

> [!NOTE]
> **Low Risk**
> Additive optional package with no changes to existing metascraper packages; errors are swallowed to preserve logo fallbacks, though adopters should configure cache TTL/store for long-running DNS/HTTP lookups.
> 
> **Overview**
> Adds a new optional **`metascraper-logo-bimi`** rules bundle that fills **`logo`** from BIMI DNS TXT records (`<selector>._bimi.<registrable-domain>`), so logos can be resolved without HTML.
> 
> The plugin parses BIMI records (single record, `v=BIMI1` first, HTTPS `l=` only, declination when `l` is empty), fetches and validates the URL as reachable **`image/svg+xml`** over HTTPS (including post-redirect checks), and **memoizes** per-domain results while **not** caching resolver/URL failures. DNS and URL checks are **injectable** (`resolveTxt`, `resolveLogoUrl`, `gotOpts`, `keyvOpts`, `selector`); failures return `undefined` so later **`metascraper-logo`** / favicon rules still run.
> 
> The root **README** official bundles list is updated. Existing behavior is unchanged unless callers add the package (typically listed before other logo rules).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 416f91e8003e786bc1f0289bc9e4cc1a55082f7f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added `metascraper-logo-bimi` support to discover and validate BIMI SVG logo URLs from DNS TXT records.
  * BIMI logos are preferred and evaluated first, with HTTPS + SVG checks and graceful fallback when BIMI is missing or invalid.
  * Added configurable selector, DNS/TXT and URL-resolution options, plus per-domain memoization.
* **Documentation**
  * Added package documentation and updated the official rules bundles list.
* **Tests**
  * Added AVA coverage for parsing BIMI TXT, selector behavior, lookup/validation edge cases, caching, and metascraper integration.
* **Refactor**
  * Added TypeScript typings for plugin options and rules entry point.

### CODE REVIEW

File: packages/metascraper-logo-bimi/src/index.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+const isSvg = ({ headers }) =>
+  headers['content-type']?.split(';')[0].trim().toLowerCase() ===
+  SVG_CONTENT_TYPE
+
+const defaultResolveLogoUrl = async (logoUrl, gotOpts) => {
+  const response = await reachableUrl(logoUrl, gotOpts)
+  return reachableUrl.isReachable(response) && isSvg(response)
+    ? response.url
+    : undefined
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/src/index.js]: Confirmed and fixed in fba6ce2b.

`parseRecord` enforces `https` on the record value, but `reachableUrl` follows redirects and `response.url` is the **final** URL, so the published scheme was not what got emitted. Reproduced before the fix:

```
published l= : http://127.0.0.1:63182/logo.svg
redirects to : http://127.0.0.1:63181/logo.svg
resolveLogoUrl returned: http://127.0.0.1:63182/logo.svg   <- pre-fix
resolveLogoUrl returned: undefined                          <- post-fix
```

The response decision is now a pure `toLogoUrl(response)`, so the downgrade is a deterministic test rather than one needing a TLS server:

```js
test("return undefined when a redirect downgrades the logo to http", t => {
  t.is(toLogoUrl(response({ url: "http://cdn.microlink.io/logo/logo.svg" })), undefined)
})
```

Plus a live-server case asserting a non-https origin is rejected end to end. 34 tests passing, and apple.com / cloudflare.com / x.com still resolve.

File: packages/metascraper-logo-bimi/src/index.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  selector = 'default'
+} = {}) => {
+  const getLogo = async domain => {
+    const logoUrl = await getRecord(domain, { resolveTxt, selector })
+    return logoUrl === undefined ? undefined : resolveLogoUrl(logoUrl, gotOpts)
+  }
+
+  const fn = memoize(getLogo, keyvOpts, {
+    value: value => (value === undefined ? null : value)
+  })
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/src/index.js:90]: Correct on the facts, declining the code change. Documented instead in fba6ce2b.

`@keyvhq/memoize` does `new Keyv(keyvOptions)` and Keyv's default store is an unbounded `Map`, so distinct domain misses are retained. But this is identical to `metascraper-logo-favicon`, which uses the same `memoize(getLogo, keyvOpts, ...)` shape and has run in production at Microlink for years. Adding a bounded default here alone would diverge from the sibling package and pull in an LRU dependency for a package that otherwise mirrors it.

There is also no first-party bounded store to reach for: `@keyvhq/max-size` caps *value* size, not entry count.

So `keyvOpts` stays the escape hatch, and the README now says so instead of leaving it implicit:

> The default store is an in-memory map that never evicts, so a long running process scraping many domains should supply its own store, plus a `ttl` in milliseconds to bound how long a record is trusted.

I traced the `ttl` path to be sure it is real advice, since `memoize` passes its own `ttl` arg through: `keyv.set(key, value, undefined)` falls back to `this.ttl` from the constructor. Verified:

```
calls after 2 immediate lookups: 1
calls after ttl expiry: 2
```

Worth doing across both logo packages at once if you want a bounded default, rather than only on the new one.

File: packages/metascraper-logo-bimi/README.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+## Install
+
+```bash
+$ npm install metascraper-logo-bimi --save
+```
+
+## Usage
+
+Rules are evaluated in the order the packages are declared, so put it first to prefer the verified logo over anything found in the markup:
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/README.md]: Fair, fixed in fba6ce2b. The package reads `l=` and never looks at `a=`, so calling it "the verified logo" overclaimed.

Usage section now reads "prefer the BIMI published logo over anything found in the markup", and the rationale says it explicitly:

> The record is published in the domain's own DNS, so it is self asserted: the same level of trust as `og:logo`. Domains may also publish a Verified Mark Certificate under `a=`, where a certificate authority has attested the mark against the trademark owner, but this package does not read or validate it, so treat the result as a BIMI published logo rather than a verified one.

Also dropped "verified" from the `gotOpts` description, where it meant reachability and read as VMC validation.

File: packages/metascraper-logo-bimi/README.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  <br>
+  <br>
+  <p align="center"><strong>metascraper-logo-bimi</strong>: Get logo property from the BIMI DNS record.</p>
+  <p align="center">See our <a href="https://metascraper.js.org" target='_blank' rel='noopener noreferrer'>website</a> for more information.</p>
+  <br>
+</div>
+
+## Why
+
+[BIMI](https://datatracker.ietf.org/doc/html/draft-blank-ietf-bimi) is the standard behind the brand logo mailbox providers show next to an email. Domains publish it as a TXT record:
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/README.md]: Half right, so half applied in fba6ce2b.

**The draft link — changed.** Verified via the datatracker API: `draft-blank-ietf-bimi` is at rev02 and expired `2021-09-10`. But it was not replaced by another core spec. The active documents are `draft-brotman-ietf-bimi-guidance` (rev15) and `draft-adams-bimi-reporting` (rev11), which are implementation guidance and reporting, not the record syntax this package parses. So there is no newer spec to point at.

Switched from the rev-pinned HTML to the datatracker document page, which shows current status rather than presenting an expired revision as live:

`https://datatracker.ietf.org/doc/html/draft-blank-ietf-bimi` -> `https://datatracker.ietf.org/doc/draft-blank-ietf-bimi/`

**The homepage 404 — not changed.** It is a 404, but it is the repo-wide convention: 28 of the packages use the same `github.com/microlinkhq/metascraper/packages/<name>` shape, including `metascraper-logo-favicon`, which also returns 404. Fixing it on the new package alone makes it the odd one out. Worth a separate sweep across every `package.json` if you want it corrected.

File: packages/metascraper-logo-bimi/package.json [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    "bimi",
+    "dns",
+    "logo",
+    "metascraper"
+  ],
+  "dependencies": {
+    "@keyvhq/memoize": "~2.2.4",
+    "@metascraper/helpers": "workspace:*",
+    "reachable-url": "~1.8.3"
+  },
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/package.json:31]: Declining, following existing precedent.

`metascraper-logo-favicon/src/index.d.ts` imports the exact same two modules:

```ts
gotOpts?: import("got").Options,
keyvOpts?: import("@keyvhq/core").Options<any>,
```

and declares neither, with dependencies `@keyvhq/memoize`, `@metascraper/helpers`, `lodash`, `reachable-url`. This package mirrors that exactly.

Adding them would have to be `dependencies`, not `devDependencies`, for consumers to resolve the types, which means shipping a runtime `got@11` to a package that never calls `require("got")`. Both are already reachable transitively: `got` via `reachable-url`, `@keyvhq/core` via `@keyvhq/memoize`.

Same call as the homepage one: if the type-resolution contract should be tightened, it belongs in a sweep over every package that references `got` in its `.d.ts`, not only the new one.

File: packages/metascraper-logo-bimi/README.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+</div>
+
+## Why
+
+[BIMI](https://datatracker.ietf.org/doc/draft-blank-ietf-bimi/) is the standard behind the brand logo mailbox providers show next to an email. Domains publish it as a TXT record:
+
+```
+$ dig +short TXT default._bimi.microlink.io
+"v=BIMI1; l=https://cdn.microlink.io/logo/logo.svg;"
+```
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/README.md]: Fixed in 90ce1945. That was the only unlabeled fence in the README, so labelling it `console` is free and does not conflict with anything else in the repo.

File: packages/metascraper-logo-bimi/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+That makes it a higher quality source than a favicon, and it doesn't need the markup: a single DNS lookup, so it works even when the page is JavaScript rendered or unreachable.
+
+Coverage is the trade-off. It's common among large brands and rare in the long tail, so pair it with [metascraper-logo](https://github.com/microlinkhq/metascraper/tree/master/packages/metascraper-logo) and [metascraper-logo-favicon](https://github.com/microlinkhq/metascraper/tree/master/packages/metascraper-logo-favicon).
+
+## Install
+
+```bash
+$ npm install metascraper-logo-bimi --save
+```
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/README.md:32]: Declining, same reasoning as the homepage finding.

All 32 package READMEs in this repo use the identical install block:

```bash
$ npm install <package> --save
```

Verified: `grep -l "^\$ npm install" packages/*/README.md | wc -l` returns 32 of 32. Dropping the `$` here alone makes this the only package that differs, for a lint rule the repo does not enforce (there is no markdownlint config or CI step for it).

Worth a repo-wide sweep together with the homepage URLs if you want MD014 clean, rather than a one-package exception.

**Comment by Kikobeats** [comment]
CI status: **58 of 59 checks pass**. The one red is `Coveralls - metascraper-video`, and it is not attributable to this branch.

**What it says:** `Coverage decreased (-0.009%) to 98.675%`.

**Why it is not this PR:** the diff outside `packages/metascraper-logo-bimi/` is exactly one line, the package-list entry in the root `README.md`. Zero lines land in `metascraper-video`.

**Why it fires anyway:** that package's coverage oscillates on its own, because all four of its test files hit live network and take different branches per run. Back-to-back commits on master:

| commit | branch | Coveralls - metascraper-video |
|---|---|---|
| `ca04c634` | master | Coverage at **98.675%** |
| `be9b049d` | master | Coverage increased (+0.009%) to **98.684%** |
| `90ce1945` | this PR | Coverage decreased (-0.009%) to **98.675%** |

The value bounces between 98.675% and 98.684%. This PR reads the same number master itself read one commit earlier; whichever side the base build landed on decides whether the next build reports "increased" or "decreased".

I re-ran `Test metascraper-video` once and it came back on the same side. I am not going to keep re-rolling CI to win a coin flip, and adding tests to an untouched package to move a 0.009% delta would be papering over the nondeterminism rather than fixing it.

Checks that are actually this PR are green: `Test metascraper-logo-bimi`, `Coveralls - metascraper-logo-bimi`. `Test metascraper-logo-favicon` also cleared on its own after an earlier flake in `resolve-favicon-url > undefined if body is not present`, which races a 100ms `res.destroy()` against a live JPEG stream from `cdn.microlink.io`.

The real fix is to stabilise the live-network suites, which is its own PR.

File: packages/metascraper-logo-bimi/src/index.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  try {
+    answers = await resolveTxt(`${selector}._bimi.${domain}`)
+  } catch {
+    return undefined
+  }
+
+  for (const answer of answers) {
+    const logoUrl = parseRecord(answer.join(''))
+    if (logoUrl) return logoUrl
+  }
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/metascraper-logo-bimi/src/index.js]: Confirmed and fixed in f2268dae.

`getRecord` scanned every TXT answer and returned the first that yielded a logo, so `parseRecord` returning `undefined` meant two different things: "not a BIMI record" and "a BIMI record that declines". The declination was skipped like an unrelated answer. Reproduced before the fix:

```
answers  : [['v=BIMI1; l=; a=;'], ['v=BIMI1; l=https://cdn.microlink.io/logo/logo.svg;']]
pre-fix  : https://cdn.microlink.io/logo/logo.svg
post-fix : undefined
```

The loop existed to skip other TXT records sharing the hostname, which is real, but the spec already separates those two steps. Section 7.2:

> Records that do not start with a "v=" tag that identifies the current version of BIMI MUST be discarded. [...] If the remaining set contains multiple records or no records, Assertion Record Discovery terminates and BIMI processing MUST NOT be performed.

So the discard is by version tag, and what survives has to be exactly one record:

```js
const records = answers.map(answer => answer.join('')).filter(isRecord)
return records.length === 1 ? parseRecord(records[0]) : undefined
```

That fixes the declination and also the case the bug did not mention: two BIMI records both publishing a logo, where the first silently won instead of the set being rejected.

Two tests added, one per branch of the rule:

```js
test('return undefined when a second BIMI record is published', ...)
test('honor a declination published next to another BIMI record', ...)
```

36 tests passing. The existing `v=spf1` coexistence test still passes, and apple.com, cloudflare.com, x.com, vercel.com and microlink.io still resolve, with github.com still `undefined`.
